File: 3.0/qqzeng-ip-php/IpSearch.php
<?php
/* 
PHP Version 7.2.2
qqzeng-ip 3.0 2018-08-18

在Java C# 中单例会一直存在于整个应用程序的生命周期里，变量是跨页面级的，真正可以做到这个实例在应用程序生命周期中的唯一性。
然而在PHP中，所有的变量无论是全局变量还是类的静态成员，都是页面级的，每次页面被执行时，都会重新建立新的对象，都会在页面执行完毕后被清空


*/

class IpSearch  {
  
    private $prefStart=array(); 
    private $prefEnd=array(); 
    private $endArr=array(); 
    private $addrArr=array(); 
    private $fp;
 
    static private   $instance= null;

    private    function __construct() { 
        $this->loadDat(); 
    }

    private  function loadDat(){
        $path='qqzeng-ip-3.0-ultimate.dat';
        $this->fp = fopen($path, 'rb');
        $fsize = filesize($path);
     
        $data = fread( $this->fp, $fsize); 

        for ($k = 0; $k < 256; $k++)
        {
            $i = $k * 8 + 4;           
            $this->prefStart[$k]=unpack("V",$data,$i)[1];
            $this->prefEnd[$k]=unpack("V",$data,$i+4)[1];               
        }

        $RecordSize =unpack("V",$data,0)[1];  

        for ($i = 0; $i <$RecordSize; $i++)
        {
            $p = 2052 + ($i * 8);         
            $this->endArr[$i] =unpack("V",$data,$p)[1];
      
            $offset =  unpack("V", $data[4 + $p]. $data[5 + $p]. $data[6 + $p]  . "\x0")[1];
            $length =unpack("C", $data[7 + $p])[1];
          
            fseek($this->fp, $offset);
            $this->addrArr[$i] =fread( $this->fp,  $length);
        }


    }

    function __destruct(){
        if ($this->fp !== NULL) {
            fclose($this->fp);
        }
    }

    private  function __clone() {}
    private  function __wakeup() {}
        
    public static function getInstance() {
        if (self::$instance instanceof IpSearch ) {           
            return self::$instance;
        }
        else{
          
            return self::$instance = new IpSearch();
          
        }
       
    }
    
    public function get($ip) {
        $val =sprintf("%u",ip2long($ip));
        $ip_arr = explode('.', $ip); 
        $pref = $ip_arr[0];
        $low = $this->prefStart[$pref];
        $high =  $this->prefEnd[$pref];
        $cur = $low == $high ? $low : $this->BinarySearch($low, $high, $val);
        return $this->addrArr[$cur];
    }

    private function BinarySearch($low, $high, $k) {
        $M = 0;
        while ($low <= $high) {
            $mid = floor(($low + $high) / 2);
            $endipNum = $this->endArr[$mid];           
            if ($endipNum >= $k) {
                $M = $mid;
                if ($mid == 0) {
                    break;
                }
                $high = $mid - 1;
            } else $low = $mid + 1;
        }
        return $M;
    }
   
    
}
 
?>

File: archive/3.0/qqzeng-ip-java/IpSearch.java
import java.io.IOException;
import java.nio.charset.StandardCharsets;
import java.nio.file.Files;
import java.nio.file.Path;
import java.nio.file.Paths;
import java.util.*;

public class IpSearch {

	private int[] prefStart = new int[256];
	private int[] prefEnd = new int[256];
	private long[] endArr;
	private String[] addrArr;

	private static IpSearch instance = null;

	private IpSearch() {
		Path path = Paths.get("C:\\qqzeng-ip-3.0-ultimate.dat");

		byte[] data = null;
		try {
			data = Files.readAllBytes(path);
		} catch (IOException e) {
			e.printStackTrace();
		}

		for (int k = 0; k < 256; k++) {
			int i = k * 8 + 4;
			prefStart[k] = (int) BytesToLong(data[i], data[i + 1], data[i + 2], data[i + 3]);
			prefEnd[k] = (int) BytesToLong(data[i + 4], data[i + 5], data[i + 6], data[i + 7]);

		}

		int RecordSize = (int) BytesToLong(data[0], data[1], data[2], data[3]);
		endArr = new long[RecordSize];
		addrArr = new String[RecordSize];
		for (int i = 0; i < RecordSize; i++) {
			int p = 2052 + (i * 8);
			long endipnum = BytesToLong(data[p], data[1 + p], data[2 + p], data[3 + p]);

			int offset = (int)BytesToLong3(data[4 + p], data[5 + p], data[6 + p]);
			int length = data[7 + p] & 0xff;

			endArr[i] = endipnum;

			addrArr[i] = new String(Arrays.copyOfRange(data,  offset, (offset + length)));
		}

	}

	public static synchronized IpSearch getInstance() {
		if (instance == null)
			instance = new IpSearch();
		return instance;
	}

	public String Get(String ip) {

		String[] ips = ip.split("\\.");
		int pref = Integer.valueOf(ips[0]);
		long val = ipToLong(ip);
		int low = prefStart[pref], high = prefEnd[pref];
		long cur = low == high ? low : BinarySearch(low, high, val);
		return addrArr[(int) cur];

	}

	private int BinarySearch(int low, int high, long k) {
		int M = 0;
		while (low <= high) {
			int mid = (low + high) / 2;

			long endipNum = endArr[mid];
			if (endipNum >= k) {
				M = mid;
				if (mid == 0) {
					break; 
				}
				high = mid - 1;
			} else
				low = mid + 1;
		}
		return M;
	}

	private long BytesToLong(byte a, byte b, byte c, byte d) {
		return (a & 0xFFL) | ((b << 8) & 0xFF00L) | ((c << 16) & 0xFF0000L) | ((d << 24) & 0xFF000000L);

	}

	private long BytesToLong3(byte a, byte b, byte c) {
		return (a & 0xFFL) | ((b << 8) & 0xFF00L) | ((c << 16) & 0xFF0000L);

	}

	private long ipToLong(String ip) {
		long result = 0;
		String[] d = ip.split("\\.");
		for (String b : d) {
			result <<= 8;
			result |= Long.parseLong(b) & 0xff;
		}
		return result;
	}

	public static void main(String[] args) {

		IpSearch finder = IpSearch.getInstance();

		String ip = "8.8.8.8";
		String result = finder.Get(ip);
		System.out.println(ip);
		System.out.println(result);

		/*
		 * 1.197.224.9 亚洲|中国|河南|周口|商水|电信|411623|China|CN|114.60604|33.53912
		 */

	}

}
